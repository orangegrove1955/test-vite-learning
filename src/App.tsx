import { useEffect, useState } from "react";
import { Input } from "./components/ui/input";
import { Button } from "./components/ui/button";
import { Checkbox } from "./components/ui/checkbox";
import {
  addTask,
  deleteTask,
  fetchTasks,
  toggleTask,
  type Task,
} from "./services/taskService";

function App() {
  const [tasks, setTasks] = useState<Task[]>([]);
  const [input, setInput] = useState("");
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    fetchTasks()
      .then(setTasks)
      .catch((err) => setError(err.message))
      .finally(() => setLoading(false));
  }, []);

  const handleAddTask = async () => {
    if (!input.trim()) return;
    try {
      const newTask = await addTask(input);
      setTasks((prev) => [...prev, newTask]);
      setInput("");
    } catch (err) {
      console.error(err);
      alert("Failed to add task");
    }
  };

  const handleToggle = async (id: number) => {
    try {
      await toggleTask(id);
      setTasks((prev) =>
        prev.map((t) => (t.id === id ? { ...t, done: !t.done } : t))
      );
    } catch {
      alert("Failed to toggle task");
    }
  };

  const handleDelete = async (id: number) => {
    try {
      await deleteTask(id);
      setTasks((prev) => prev.filter((t) => t.id !== id));
    } catch {
      alert("Failed to delete task");
    }
  };

  if (loading) return <p className="p-4">Loading tasks...</p>;
  if (error) return <p className="p-4 text-red-500">Error: {error}</p>;

  return (
    <div className="max-w-md mx-auto mt-10 p-4 space-y-4">
      <h1 className="text-2xl font-bold">Task Tracker Lite</h1>
      <div className="flex gap-2">
        <Input
          value={input}
          onChange={(e) => setInput(e.target.value)}
          placeholder="Add a task"
        />
        <Button onClick={handleAddTask}>Add</Button>
      </div>
      <ul className="space-y-2">
        {tasks.map((task) => (
          <li
            key={task.id}
            className="flex justify-between items-center p-3 border rounded"
          >
            <div
              className="flex gap-2 items-center  cursor-pointer"
              onClick={() => handleToggle(task.id)}
            >
              <Checkbox id={`check-${task.id}`} checked={task.done} />
              <label htmlFor={`check-${task.id}`}>
                <span
                  className={`${task.done ? "line-through text-gray-500" : ""}`}
                >
                  {task.text}
                </span>
              </label>
            </div>

            <Button variant="destructive" onClick={() => handleDelete(task.id)}>
              X
            </Button>
          </li>
        ))}
      </ul>
    </div>
  );
}

export default App;
