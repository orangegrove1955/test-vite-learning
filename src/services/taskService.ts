export type Task = { id: number; text: string; done: boolean };

const BASE_URL = "https://example.com/api/tasks"; // placeholder

export const fetchTasks = async (): Promise<Task[]> => {
  const res = await fetch(BASE_URL);
  return res.json();
};

export const addTask = async (text: string): Promise<Task> => {
  const res = await fetch(BASE_URL, {
    method: "POST",
    body: JSON.stringify({ text, done: false }),
    headers: { "Content-Type": "application/json" },
  });
  return res.json();
};

export const toggleTask = async (id: number): Promise<void> => {
  await fetch(`${BASE_URL}/${id}/toggle`, { method: "POST" });
};

export const deleteTask = async (id: number): Promise<void> => {
  await fetch(`${BASE_URL}/${id}`, { method: "DELETE" });
};
