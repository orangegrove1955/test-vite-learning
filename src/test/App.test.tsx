import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import App from "../App";
import * as taskService from "../services/taskService";

describe("App Component", () => {
  const mockTasks = [
    { id: 1, text: "Test Task 1", done: false },
    { id: 2, text: "Test Task 2", done: true },
  ];

  beforeEach(() => {
    // Reset all mocks before each test
    vi.resetAllMocks();

    // Use spyOn to mock the implementations of the taskService functions
    vi.spyOn(taskService, "fetchTasks").mockResolvedValue([...mockTasks]);
    vi.spyOn(taskService, "addTask").mockImplementation(async (text) => ({
      id: Math.floor(Math.random() * 1000),
      text,
      done: false,
    }));
    vi.spyOn(taskService, "toggleTask").mockResolvedValue(undefined);
    vi.spyOn(taskService, "deleteTask").mockResolvedValue(undefined);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("renders the app title", async () => {
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("Task Tracker Lite")).toBeInTheDocument();
    });
  });

  it("shows loading state initially", () => {
    render(<App />);
    expect(screen.getByText("Loading tasks...")).toBeInTheDocument();
  });

  it("displays tasks from API", async () => {
    render(<App />);

    await waitFor(() => {
      expect(taskService.fetchTasks).toHaveBeenCalledTimes(1);
      expect(screen.getByText("Test Task 1")).toBeInTheDocument();
      expect(screen.getByText("Test Task 2")).toBeInTheDocument();
    });
  });

  it("handles API error", async () => {
    // Override the mock implementation just for this test
    vi.spyOn(taskService, "fetchTasks").mockRejectedValueOnce(
      new Error("Failed to fetch")
    );

    render(<App />);

    await waitFor(() => {
      expect(screen.getByText(/Error: Failed to fetch/i)).toBeInTheDocument();
    });
  });

  it("adds a new task", async () => {
    vi.spyOn(taskService, "addTask").mockResolvedValueOnce({
      id: 3,
      text: "New Task",
      done: false,
    });

    render(<App />);

    // Wait for initial tasks to load
    await waitFor(() => {
      expect(screen.getByText("Test Task 1")).toBeInTheDocument();
    });

    // Add a new task
    const input = screen.getByPlaceholderText("Add a task");
    const addButton = screen.getByText("Add");

    fireEvent.change(input, { target: { value: "New Task" } });
    fireEvent.click(addButton);

    await waitFor(() => {
      expect(taskService.addTask).toHaveBeenCalledWith("New Task");
      expect(screen.getByText("New Task")).toBeInTheDocument();
    });
  });

  it("toggles a task", async () => {
    render(<App />);

    // Wait for initial tasks to load
    await waitFor(() => {
      expect(screen.getByText("Test Task 1")).toBeInTheDocument();
    });

    // Find the first task and toggle it
    const taskItem = screen.getByText("Test Task 1").closest("div");
    expect(taskItem).not.toBeNull();

    fireEvent.click(taskItem as HTMLElement);

    await waitFor(() => {
      expect(taskService.toggleTask).toHaveBeenCalledWith(1);
    });
  });

  it("deletes a task", async () => {
    render(<App />);

    // Wait for initial tasks to load
    await waitFor(() => {
      expect(screen.getByText("Test Task 1")).toBeInTheDocument();
    });

    // Find the delete button for the first task
    const deleteButtons = screen.getAllByText("X");
    fireEvent.click(deleteButtons[0]);

    await waitFor(() => {
      expect(taskService.deleteTask).toHaveBeenCalledWith(1);
    });
  });

  it("matches snapshot", async () => {
    const { asFragment } = render(<App />);
    await screen.findByText(/Add/); // Wait until loading passes
    expect(asFragment()).toMatchSnapshot();
  });
});
