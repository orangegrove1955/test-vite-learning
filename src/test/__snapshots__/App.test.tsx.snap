// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-md mx-auto mt-10 p-4 space-y-4"
  >
    <h1
      class="text-2xl font-bold"
    >
      Task Tracker Lite
    </h1>
    <div
      class="flex gap-2"
    >
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
        data-slot="input"
        placeholder="Add a task"
        value=""
      />
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
        data-slot="button"
      >
        Add
      </button>
    </div>
    <ul
      class="space-y-2"
    >
      <li
        class="flex justify-between items-center p-3 border rounded"
      >
        <div
          class="flex gap-2 items-center  cursor-pointer"
        >
          <button
            aria-checked="false"
            class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
            data-slot="checkbox"
            data-state="unchecked"
            id="check-1"
            role="checkbox"
            type="button"
            value="on"
          />
          <label
            for="check-1"
          >
            <span
              class=""
            >
              Test Task 1
            </span>
          </label>
        </div>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 h-9 px-4 py-2 has-[>svg]:px-3"
          data-slot="button"
        >
          X
        </button>
      </li>
      <li
        class="flex justify-between items-center p-3 border rounded"
      >
        <div
          class="flex gap-2 items-center  cursor-pointer"
        >
          <button
            aria-checked="true"
            class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
            data-slot="checkbox"
            data-state="checked"
            id="check-2"
            role="checkbox"
            type="button"
            value="on"
          >
            <span
              class="flex items-center justify-center text-current transition-none"
              data-slot="checkbox-indicator"
              data-state="checked"
              style="pointer-events: none;"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-check size-3.5"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 6 9 17l-5-5"
                />
              </svg>
            </span>
          </button>
          <label
            for="check-2"
          >
            <span
              class="line-through text-gray-500"
            >
              Test Task 2
            </span>
          </label>
        </div>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 h-9 px-4 py-2 has-[>svg]:px-3"
          data-slot="button"
        >
          X
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
